### ISSUE TITLE
Hilt and ViewModel initialization with Navigation Component arguments

### ISSUE THREAD

**Issue by langsmith** [airbnb/mavericks#692]
_tldr; How do I initialize a Mavericks State with Navigation Component arguments when the fragment's ViewModel is using [`hiltMavericksViewModelFactory`](https://github.com/airbnb/mavericks/blob/main/mvrx-hilt/src/main/kotlin/com/airbnb/mvrx/hilt/HiltMavericksViewModelFactory.kt) ?_

I'm trying to figure out how to blend the use of this Mavericks library (version 3.0.6), [Hilt](https://developer.android.com/training/dependency-injection/hilt-android), and [Navigation Component](https://developer.android.com/guide/navigation/get-started) arguments. I'm not using Compose. I'm using Views. My app is single activity, multiple fragments.

I've also seen the README and various example files within [https://github.com/airbnb/mavericks/tree/main/sample-hilt](https://github.com/airbnb/mavericks/tree/main/sample-hilt).

----------------

My Gradle setup:

```
val mavericksVersion = "3.0.6"
implementation ("com.airbnb.android:mavericks:$mavericksVersion")
implementation ("com.airbnb.android:mavericks-testing:$mavericksVersion")
implementation ("com.airbnb.android:mavericks-navigation:$mavericksVersion")
    
val hiltVersion = "2.48"
implementation ("com.google.dagger:hilt-android:$hiltVersion")
kapt ("com.google.dagger:hilt-compiler:$hiltVersion")
implementation("androidx.hilt:hilt-navigation-fragment:1.0.0")    
```

----------------

`HomeState` in the `HomeViewModel` is:

```kotlin
data class HomeState(
    val userFullName: String = "",
    val paymentList: Async<List<Payment>> = Uninitialized,
    val coarseLocationPermissionGranted: Boolean = false,
    val deviceLastLocationCoordinates: Pair<Double, Double> = Pair(0.0, 0.0),
    val count: Int = 0
) : MavericksState
```

<img width="654" alt="Screenshot 2023-09-05 at 10 02 38 AM" src="https://github.com/airbnb/mavericks/assets/4394910/2e15fc1c-cc09-437e-aa4d-95b5c091ea3d">

----------------

I had Mavericks working just fine before I decided to try to install Hilt. 

```kotlin
companion object : MavericksViewModelFactory<HomeFragmentViewModel, HomeState> {

        override fun initialState(viewModelContext: ViewModelContext): HomeState {
            val homeFragment: HomeFragment = (viewModelContext as FragmentViewModelContext).fragment()
            val userFullName = homeFragment.arguments?.getString(LoginFragment.SESSION_FULL_NAME_KEY).orEmpty()
            return HomeState(userFullName = userFullName)
        }

        override fun create(viewModelContext: ViewModelContext, state: HomeState): HomeFragmentViewModel {
            return HomeFragmentViewModel(state)
        }
    }
```    

<img width="908" alt="Screenshot 2023-09-04 at 10 36 16 AM" src="https://github.com/airbnb/mavericks/assets/4394910/ae64a69c-bb10-40e6-9f0b-817d9df3054f">

----------------

My setup is the following after I dove into setting up Hilt. It works but I'm not initializing the `ViewModel` with any Navigation Component arguments: 

```kotlin
@AssistedFactory interface Factory : AssistedViewModelFactory<HomeFragmentViewModel, HomeState> {
        override fun create(state: HomeState): HomeFragmentViewModel
    }

    /* 
    companion object : MavericksViewModelFactory<HomeFragmentViewModel, HomeState> {
        override fun initialState(viewModelContext: ViewModelContext): HomeState {
            val homeFragment: HomeFragment = (viewModelContext as FragmentViewModelContext).fragment()
            val userFullName = homeFragment.arguments?.getString(LoginFragment.SESSION_FULL_NAME_KEY).orEmpty()
            return HomeState(userFullName = userFullName)
        }
    }
    */

    companion object : MavericksViewModelFactory<HomeFragmentViewModel, HomeState> by hiltMavericksViewModelFactory()
```    

<img width="1002" alt="Screenshot 2023-09-04 at 10 35 37 AM" src="https://github.com/airbnb/mavericks/assets/4394910/b02a4923-4612-44bc-b155-4f108ec48bee">

----------------

As expected, I get the crash below if I experiment by commenting out the `hiltMavericksViewModelFactory()` `companion object` setup

<img width="1231" alt="Screenshot 2023-09-04 at 10 48 57 AM" src="https://github.com/airbnb/mavericks/assets/4394910/668e8846-a151-4d46-8b15-e3e1e219bceb">

<img width="989" alt="Screenshot 2023-09-05 at 10 08 26 AM" src="https://github.com/airbnb/mavericks/assets/4394910/dbb66a31-3f7a-450e-b0b4-3004d82169dc">

----------------

Ignoring the crash for a moment, my overall question again is how I would initialize the `HomeState` with that initial `userFullName` value from the Navigation Component argument bundle? 

I see mention of `viewModelContext` in the `HiltMavericksViewModelFactory` class, so do I need to somehow adjust that `HiltMavericksViewModelFactory` class?

I haven't tried it, but would a hack be to pass the arguments (bundle) from the fragment to the `ViewModel` in the fragment's `onCreate()` or `onCreateView()`?

**Comment by langsmith** [comment]
I'm looking into https://airbnb.io/mavericks/#/fragment-arguments?id=using-fragment-args-in-the-initial-value-for-mavericksstate

The following might work:

```kotlin
data class HomeState(
    val userFullName: String,
    val paymentList: Async<List<Payment>> = Uninitialized,
    val coarseLocationPermissionGranted: Boolean = false,
    val deviceLastLocationCoordinates: Pair<Double, Double> = Pair(0.0, 0.0),
    val count: Int = 0
) : MavericksState {

constructor(homeFragmentArgs: HomeFragmentArgs) : this(userFullName = homeFragmentArgs.getString(SESSION_FULL_NAME_KEY).orEmpty())

```

**Comment by langsmith** [comment]
> I'm looking into https://airbnb.io/mavericks/#/fragment-arguments?id=using-fragment-args-in-the-initial-value-for-mavericksstate
> 
> The following might work:
> 
> ```kotlin
> LoginFragmentDirections.actionLoginFragmentToHome(userFullName = it.userFullName)
> ```
> 
> ```kotlin
> data class HomeState(
>     val userFullName: String,
>     val paymentList: Async<List<Payment>> = Uninitialized,
>     val coarseLocationPermissionGranted: Boolean = false,
>     val deviceLastLocationCoordinates: Pair<Double, Double> = Pair(0.0, 0.0),
>     val count: Int = 0
> ) : MavericksState {
> 
> constructor(homeFragmentArgs: HomeFragmentArgs) : this(userFullName = homeFragmentArgs.userFullName.orEmpty())
> ```

Nah, that alternate `constructor` setup didn't work. Not sure whether or not it's because I'm _not_ doing [the steps](https://airbnb.io/mavericks/#/fragment-arguments?id=fragment-arguments) 👇🏽 


<img width="575" alt="Screenshot 2023-09-05 at 2 03 42 PM" src="https://github.com/airbnb/mavericks/assets/4394910/720ee641-6826-4cf6-a2e9-b113e22ab1a6">

**Comment by langsmith** [comment]
Well things ARE working when I follow the steps above

1. <img width="483" alt="Screenshot 2023-09-05 at 2 16 40 PM" src="https://github.com/airbnb/mavericks/assets/4394910/623447af-1c88-4715-8798-09b0654a0544">

2. <img width="809" alt="Screenshot 2023-09-05 at 2 16 43 PM" src="https://github.com/airbnb/mavericks/assets/4394910/076018db-6a13-4875-bbfe-bd319f1d2cfa">

3. <img width="803" alt="Screenshot 2023-09-05 at 2 16 52 PM" src="https://github.com/airbnb/mavericks/assets/4394910/4eb426da-cc17-44ca-9968-fa31516aeffb">

However, this process seems opposite to using the arguments declared in the Navigation Component nav graph XML file. I still have to create custom Parcelizable argument data classes 😕  I'm unable to use generated `_____Directions` and do something like

```kotlin
findNavController().navigate(LoginFragmentDirections.actionLoginFragmentToHome(userFullName = it.userFullName))
```


I know that [custom `Parcelable` objects **can** be passed through as arguments](https://developer.android.com/guide/navigation/use-graph/pass-data)

<img width="853" alt="Screenshot 2023-09-05 at 2 58 31 PM" src="https://github.com/airbnb/mavericks/assets/4394910/ddb5564e-f9fd-4129-a657-fff3fb399885">

------------------------------------------------------------------------------------------------

However, this doesn't really help me from avoiding having to create these custom args classes in the first place 😕

**Comment by langsmith** [comment]
Any thoughts on this @gpeal @elihart ?

**Comment by langsmith** [comment]
Doesn't really address this ticket's topic/question, but another option I just thought of was to save the full name to repo-level persistence in `LoginViewModel` and just fetch it in `HomeViewModel` instead of passing it at Navigation Component arguments from one fragment/VM to another.

**Comment by langsmith** [comment]
Closing the loop here. This is what I've ultimately ended up with and I'm posting it here mainly so that others might benefit from it. 

I created a `setAsMavericksArgs` extension function, which has made things pretty easy and minimal when going from one fragment/VM to another.

```kotlin
import android.os.Parcelable
import kotlinx.parcelize.Parcelize

@Parcelize
class ParcelableClassWithArgs(
    var stringValue: String? = "",
    var booleanValue: Boolean? = false,
): Parcelable
```


`Parcelable` extension function 👇🏽 

```kotlin
fun Parcelable.setAsMavericksArgs(fragment: Fragment): Bundle {
    this.asMavericksArgs().let {
        fragment.arguments = it
        return it
    }
}
```

`Fragment` extension function 👇🏽 

```kotlin
fun Fragment.navigate(actionInt: Int, optionalArgs: Bundle? = null) =
    findNavController().navigate(actionInt, optionalArgs)
```

Navigating in the `Fragment` 👇🏽 

```kotlin
navigate(
    NavGraphDirections.actionToNextFragment().actionId,
    ParcelableClassWithArgs(
		...various values
	).setAsMavericksArgs(this)
)
```    

Setup in the `ViewModel` of the `Fragment` that has been navigated _to_
```kotlin
data class ReceivingViewModelState(
    val args: ParcelableClassWithArgs? = null,
) : MavericksState {
    constructor(parcelableClassWithArgs: ParcelableClassWithArgs) : this(args = parcelableClassWithArgs)
}
```
